src/main/java/net/imglib2/type/numeric/integer/IntegerTypes.java

```diff
@@ -0,0 +1,42 @@
+package net.imglib2.type.numeric.integer;
+
+import net.imglib2.type.logic.BitType;
+import net.imglib2.type.numeric.IntegerType;
+
+public class IntegerTypes
+{
+	private IntegerTypes()
+	{
+		// NB: prevent instantiation of static utility class
+	}
+
+	public static IntegerType< ? > smallestType( boolean signed, long max )
+	{
+		return smallestType( signed ? -1 : 0, max );
+	}
+
+	public static IntegerType< ? > smallestType( long min, long max )
+	{
+		if ( min > max )
+			throw new IllegalArgumentException( "Wrong usage: min (" + min + ") > max (" + max + ")" );
+		if ( min >= 0 )
+		{
+			if ( max <= 1L )
+				return new BitType();
+			if ( max <= 0xffL )
+				return new UnsignedByteType();
+			if ( max <= 0xffffL )
+				return new UnsignedShortType();
+			if ( max <= 0xffffffffL )
+				return new UnsignedIntType();
+			return new UnsignedLongType();
+		}
+		if ( min >= Byte.MIN_VALUE && max <= Byte.MAX_VALUE )
```

Comment:
It'd be nice to use the `getMaxValue` and `getMinValue` methods for the bounds checks because the actual values are an implementation detail that is provided through the interface.

Reply:
Any suggestions how this should be done? In order to check `getMinValue()` and `getMaxValue()`, I first have to create an instance of the respective type, right?
Isn't it the implicit contract of e.g. `ByteType` that it can take values between `Byte.MIN_VALUE` and `Byte.MAX_VALUE`?

Reply:
> In order to check `getMinValue()` and `getMaxValue()`, I first have to create an instance of the respective type, right?

Yes, that is correct. I would create a `private static List<? extends IntegerType< ? > >` (one for signed and one for unsigned) sorted by size in ascending order, and then filter for the first element that matches the bounds, or `throw` as you do if there are none, e.g.

```java
signedIntegers
    .stream()
    .filter(i -> v >= i.getMinValue() & v <= i.getMaxValue())
    .findFirst()
    .get() // check and exception would have to come here
    .createVariable()
```

> Isn't it the implicit contract of e.g. `ByteType` that it can take values between `Byte.MIN_VALUE` and `Byte.MAX_VALUE`?

Yes it is but I personally prefer explicitly using the interface methods.

Reply:
Thanks a lot for the suggestion, @hanslovsky.

I implemented it in https://github.com/imglib/imglib2/pull/298/commits/5505beb8c78a059f3c5ac3c66d1b31efb8b45f9a. As we only allow `long` (not e.g. `BigInteger`) for `min` and `max` values, I didn't find a condition in which no type would get selected, so I skipped checking for an empty list and throwing an exception.